module/documents/actor/actor.mjs

```diff
@@ -2504,25 +2517,29 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
   /* -------------------------------------------- */
 
   /**
-   * Recovers spell slots and pact slots.
+   * Recovers expended spell slots.
    * @param {object} [options]
-   * @param {boolean} [options.recoverPact=true]     Recover all expended pact slots.
-   * @param {boolean} [options.recoverSpells=true]   Recover all expended spell slots.
+   * @param {boolean} [options.recoverShort=true]    Recover slots that return on short rests.
+   * @param {boolean} [options.recoverLong=true]     Recover slots that return on long rests.
    * @returns {object}                               Updates to the actor.
    * @protected
    */
-  _getRestSpellRecovery({recoverPact=true, recoverSpells=true}={}) {
-    const spells = this.system.spells ?? {};
+  _getRestSpellRecovery({recoverShort=true, recoverLong=true}={}) {
+    const spells = this.system.spells;
     let updates = {};
-    if ( recoverPact ) {
-      const pact = spells.pact;
-      updates["system.spells.pact.value"] = pact.override || pact.max;
-    }
-    if ( recoverSpells ) {
-      for ( let [k, v] of Object.entries(spells) ) {
-        updates[`system.spells.${k}.value`] = Number.isNumeric(v.override) ? v.override : (v.max ?? 0);
+    if ( !spells ) return updates;
+
+    Object.entries(CONFIG.DND5E.spellPreparationModes).forEach(([k, v]) => {
+      if ( v.upcast && ((recoverShort && v.shortRest) || recoverLong) ) {
+        if ( k === "prepared" ) {
+          Array.fromRange(Object.keys(CONFIG.DND5E.spellLevels).length - 1, 1).forEach(level => {
+            updates[`system.spells.spell${level}.value`] = spells[`spell${level}`].max;
+          });
```

Comment:
Unclear why we prefer to loop over the configured `spellLevels` rather than the `spells` entries. If there is no `spells` entry corresponding to an entry in `spellLevels`, won't the ``spells[`spell${level}`].max`` be undefined?

Reply:
Other way around actually, if you at any point adjust the spellLevels config object up to, say, 20 levels, then down again later, the actors will all have these properties still in the final data. For example, we ensure throughout the code that `levels` is a property in the `spells` entries to know that this is a valid one added by the system, module, or world script.

It's probably cleaner and simpler to loop throughout the `spells` object for keys of the proper name syntax and ensuring there is a `levels` or `max` property in them, however.